File: src/lib/audit.ts
import { query } from './db.js';
import { v4 as uuidv4 } from 'uuid';

export interface AuditEntry {
  agentName: string;
  actionType: string;
  riskClass?: string;
  toolName?: string;
  inputSummary?: string;
  outputSummary?: string;
  decision?: string;
  modelUsed?: string;
  tokensUsed?: number;
  costUsd?: number;
  durationMs?: number;
  error?: string;
  metadata?: Record<string, any>;
}

export async function logAudit(entry: AuditEntry, traceId?: string): Promise<void> {
  const trace = traceId || uuidv4();
  await query(
    `INSERT INTO audit_events
     (trace_id, agent_name, action_type, risk_class, tool_name, input_summary, output_summary, decision, model_used, tokens_used, cost_usd, duration_ms, error, metadata)
     VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14)`,
    [
      trace,
      entry.agentName,
      entry.actionType,
      entry.riskClass || 'R0',
      entry.toolName || null,
      entry.inputSummary || null,
      entry.outputSummary || null,
      entry.decision || 'auto_approved',
      entry.modelUsed || null,
      entry.tokensUsed || null,
      entry.costUsd || null,
      entry.durationMs || null,
      entry.error || null,
      entry.metadata ? JSON.stringify(entry.metadata) : '{}',
    ],
  );
}

File: src/lib/audit.py
"""Audit event logging — every AI call gets recorded."""

import uuid
from decimal import Decimal

from sqlalchemy import insert
from sqlalchemy.ext.asyncio import AsyncSession

from src.models import AuditLog


async def log_event(
    session: AsyncSession,
    *,
    agent: str,
    action: str,
    model_used: str | None = None,
    input_summary: str | None = None,
    output_summary: str | None = None,
    tokens_used: int | None = None,
    cost_usd: Decimal | None = None,
    duration_ms: int | None = None,
    error: str | None = None,
    trace_id: uuid.UUID | None = None,
    metadata: dict | None = None,
) -> None:
    """Insert an audit log entry."""
    await session.execute(
        insert(AuditLog).values(
            trace_id=trace_id or uuid.uuid4(),
            agent=agent,
            action=action,
            model_used=model_used,
            input_summary=input_summary[:500] if input_summary else None,
            output_summary=output_summary[:500] if output_summary else None,
            tokens_used=tokens_used,
            cost_usd=cost_usd,
            duration_ms=duration_ms,
            error=error,
            metadata_=metadata or {},
        )
    )
    await session.commit()
